katsdpcontroller/product_config.py

```diff
@@ -530,11 +530,14 @@ def __init__(
                 raise ValueError(
                     'Inconsistent centre frequencies '
                     f'(both {first.centre_frequency} and {src.centre_frequency})')
-        # TODO: this may need refinement as katxgpu evolves. For now, assume
-        # the same relation as the MeerKAT correlator i.e. 4 X-engines per
-        # F-engine, rounded up to a power of two.
+        # Determine how fine to divide the stream, i.e., the number of xgpu
+        # processes to run. The constant is the data rate (in bytes per
+        # second) that one engine can handle, and is sufficient for S-band
+        # with 80 antennas to be handled by 64 engines.
+        # The minimum is 4 since SDP expects to run 4 ingest processes.
+        first.adc_sample_rate
         n_substreams = 4
-        while n_substreams < 4 * (len(src_streams) // 2):
+        while n_substreams * 4.375e9 < first.adc_sample_rate * len(src_streams):
```

Comment:
I'd just suggest making this more explicit by giving this constant a name:
```
max_datarate_per_engine: Final = 4.375e9
```
I know you've referred to it in the comment, but I can imagine the number (or even the comparison) being modified without the comment being updated.

Reply:
Good point - I've added a constant to defaults.py where some similar constants live.